browser/ui/webui/new_tab_takeover/android/new_tab_takeover_ui.cc

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/ui/webui/new_tab_takeover/android/new_tab_takeover_ui.h"
+
+#include <memory>
+#include <ranges>
+#include <string>
+#include <utility>
+
+#include "base/strings/stringprintf.h"
+#include "base/values.h"
+#include "brave/browser/ui/webui/brave_webui_source.h"
+#include "brave/components/brave_new_tab_ui/new_tab_takeover/grit/new_tab_takeover_generated_map.h"
+#include "brave/components/constants/webui_url_constants.h"
+#include "brave/components/ntp_background_images/browser/ntp_sponsored_rich_media_ad_event_handler.h"
+#include "brave/components/ntp_background_images/browser/url_constants.h"
+#include "brave/components/ntp_background_images/browser/view_counter_service.h"
+#include "chrome/browser/ui/android/tab_model/tab_model.h"
+#include "chrome/browser/ui/android/tab_model/tab_model_list.h"
+#include "components/grit/brave_components_resources.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_ui.h"
+#include "content/public/browser/web_ui_data_source.h"
+#include "content/public/common/url_constants.h"
+
+namespace {
+
+content::WebContents* GetActiveWebContents() {
```

Comment:
is this webui not displayed in a tab? Why do we need to lookup the webcontents?

Reply:
The reason is that  the current New Tab Takeover web contents is displayed in the ThinWebView so it is not connected to the Android Tab, i.e. WebContents::GetDelegate() returns `nullptr`. Therefore to do a Tab navigation, we need to locate the current Android Tab and open the URL in it.

I left a comment in the code here: https://github.com/brave/brave-core/pull/27645/files#diff-1985d9c811d8596c2b92edcd5d8c4861e99dc76e48d07db05807d9221e81899fR142

Reply:
cc @SergeyZhukovsky and @deeppandya to confirm